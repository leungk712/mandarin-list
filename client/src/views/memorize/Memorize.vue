<template>
  <div>
    <v-container class="container-border memorize-container mb-6">
      <h1 class="display-1 my-4">Memorize!</h1>
      <p>Test your vocabulary by hovering over a card to see the meaning!</p>
      <v-divider />
      <v-row>
        <v-col
            v-for="(character) in charactersList"
            :key="character._id"
            sm="6"
            md="4"
            lg="3"
            xl="2"
        >
        <div class="memorize-card">
            <div class="memorize-card-inner">
                <v-card
                    class="my-2 pa-2 memorize-card-front"
                    outlined
                    tag="div"
                    hover
                >
                    <v-card-text>
                        <v-row justify="center" align="center">
                            <v-col cols="12" class="mt-10">
                                <p class="display-2 text--primary">{{ character.character }}</p>
                            </v-col>
                        </v-row>
                    </v-card-text>
                </v-card>
                <v-card
                    class="mx-auto my-2 pa-2 memorize-card-back"
                    outlined
                    tag="div"
                    hover
                >
                    <v-card-text>
                        <v-row justify="center" align="center">
                            <v-col cols="12" class="ma-0 pa-0">
                                <p class="title">{{ character.pinyin }}</p>
                            </v-col>
                            <v-col cols="12" class="ma-0 pa-0">
                                <h1 class="display-2">{{ character.character }}</h1>
                            </v-col>
                            <v-col cols="12" class="mt-3">
                                <p class="title">{{ character.english }}</p>    
                            </v-col>
                        </v-row>
                    </v-card-text>
                </v-card>
            </div>
        </div>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import { namespace, State } from "vuex-class";
import { PostsState, SelectedCharacter } from "@/models";
import PostsModule from "@/store/modules/posts";
const posts = namespace(PostsModule.name);

@Component({
  name: "Memorize"
})
export default class Memorize extends Vue {
  // ===== Store ===== //
  @State("posts") public posts!: PostsState;

  // ===== Data ===== //

  // ===== Methods ===== //

  // ===== Computed ===== //
get charactersList(): SelectedCharacter[] {
    return this.posts.mandarinList.sort(() => 0.5 - Math.random());
  }

  // ===== Lifecycle Hooks ===== //
}
</script>

<style scoped>
    .memorize-container {
        width: 100%;
    }
    .memorize-card {
        height: 200px;
        width: 200px;
    }

    .memorize-card-inner {
        position: relative;
        width: 100%;
        height: 100%;
        text-align: center;
        transition: transform 0.8s;
        transform-style: preserve-3d;
    }

    .memorize-card:hover .memorize-card-inner {
        transform: rotateY(180deg);
    }

    .memorize-card-front, .memorize-card-back {
        position: absolute;
        width: 100%;
        height: 100%;
        backface-visibility: hidden;
        background-color: #D1C4E9;
    }

    .memorize-card-back {
        color: white;
        transform: rotateY(180deg);
    }
</style>
