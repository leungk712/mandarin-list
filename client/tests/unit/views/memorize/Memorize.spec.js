import { createLocalVue, shallowMount } from "@vue/test-utils";
import Memorize from "@/views/memorize/Memorize.vue";
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";
import { mockMandarinList } from "../../mocks/mock-mandarin-list";
import { mockCategoriesList } from "../../mocks/mock-categories-list";

Vue.use(Vuetify);

const localVue = createLocalVue();
localVue.use(Vuex);

const elem = document.createElement("div");
if (document.body) {
    document.body.appendChild(elem);
};

document.body.setAttribute("data-app", true);

const actions = {};

const mockStore = {
    categories: {
        state: {
            categoriesList: mockCategoriesList
        },
        actions
    },
    posts: {
        state: {
            mandarinList: mockMandarinList
        },
        actions
    }
};

describe('Memorize', () => {
    let wrapper;
    const store = new Vuex.Store({
        modules: mockStore
    });
    const mountFunction = (options, store = mockStore) => {
        return shallowMount(Memorize, {
           attachTo: elem,
           localVue,
           store,
           vuetify: new Vuetify(),
           ...options
        });
    };
    describe("template", () => {
        it("should match snapshot", () => {
            wrapper = mountFunction({}, store);
            expect(wrapper.vm.$el).toMatchSnapshot();
            wrapper.destroy();
        });
    });
});