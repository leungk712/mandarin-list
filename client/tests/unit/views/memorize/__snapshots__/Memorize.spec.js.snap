// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Memorize template should match snapshot 1`] = `
<div>
  <v-container-stub
    class="container-border memorize-container mb-6"
    tag="div"
  >
    <h1
      class="display-1 my-4"
    >
      Memorize!
    </h1>
     
    <p>
      Test your vocabulary by hovering over a card to see the meaning!
    </p>
     
    <v-divider-stub
      class="my-2"
    />
     
    <v-row-stub
      class="mt-1"
      tag="div"
    >
      <v-col-stub
        cols="6"
        tag="div"
      >
        <v-select-stub
          appendicon="$dropdown"
          backgroundcolor=""
          chips="true"
          class="memorize-category-menu my-0 py-0"
          clearicon="$clear"
          data-testid="memorize-categories-menu"
          errorcount="1"
          errormessages=""
          itemcolor="primary"
          itemdisabled="disabled"
          items=""
          itemtext="name"
          itemvalue="value"
          label="Select one or more categories to view the associated characters"
          loaderheight="2"
          menuprops="[object Object]"
          messages=""
          multiple="true"
          nodatatext="$vuetify.noDataText"
          outlined="true"
          rules=""
          successmessages=""
          type="text"
          value=""
          valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date && a.getTime() !== b.getTime()) {
    // If the values are Date, compare them as timestamps
    return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
        />
      </v-col-stub>
    </v-row-stub>
     
    <v-row-stub
      tag="div"
    >
      <v-col-stub
        lg="3"
        md="4"
        sm="6"
        tag="div"
        xl="2"
      >
        <div
          class="memorize-card"
        >
          <div
            class="memorize-card-inner"
          >
            <v-card-stub
              class="my-2 pa-2 memorize-card-front"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="mt-10"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="display-2 text--primary"
                    >
                      痛
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
             
            <v-card-stub
              class="mx-auto my-2 pa-2 memorize-card-back"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Tòng
                    </p>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <h1
                      class="display-2"
                    >
                      痛
                    </h1>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="mt-3"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Pain
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
          </div>
        </div>
      </v-col-stub>
      <v-col-stub
        lg="3"
        md="4"
        sm="6"
        tag="div"
        xl="2"
      >
        <div
          class="memorize-card"
        >
          <div
            class="memorize-card-inner"
          >
            <v-card-stub
              class="my-2 pa-2 memorize-card-front"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="mt-10"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="display-2 text--primary"
                    >
                      再见
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
             
            <v-card-stub
              class="mx-auto my-2 pa-2 memorize-card-back"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Zàijiàn
                    </p>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <h1
                      class="display-2"
                    >
                      再见
                    </h1>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="mt-3"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Goodbye
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
          </div>
        </div>
      </v-col-stub>
      <v-col-stub
        lg="3"
        md="4"
        sm="6"
        tag="div"
        xl="2"
      >
        <div
          class="memorize-card"
        >
          <div
            class="memorize-card-inner"
          >
            <v-card-stub
              class="my-2 pa-2 memorize-card-front"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="mt-10"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="display-2 text--primary"
                    >
                      咳嗽
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
             
            <v-card-stub
              class="mx-auto my-2 pa-2 memorize-card-back"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Késòu
                    </p>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <h1
                      class="display-2"
                    >
                      咳嗽
                    </h1>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="mt-3"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Cough
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
          </div>
        </div>
      </v-col-stub>
      <v-col-stub
        lg="3"
        md="4"
        sm="6"
        tag="div"
        xl="2"
      >
        <div
          class="memorize-card"
        >
          <div
            class="memorize-card-inner"
          >
            <v-card-stub
              class="my-2 pa-2 memorize-card-front"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="mt-10"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="display-2 text--primary"
                    >
                      山
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
             
            <v-card-stub
              class="mx-auto my-2 pa-2 memorize-card-back"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Shān
                    </p>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <h1
                      class="display-2"
                    >
                      山
                    </h1>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="mt-3"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Mountain
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
          </div>
        </div>
      </v-col-stub>
      <v-col-stub
        lg="3"
        md="4"
        sm="6"
        tag="div"
        xl="2"
      >
        <div
          class="memorize-card"
        >
          <div
            class="memorize-card-inner"
          >
            <v-card-stub
              class="my-2 pa-2 memorize-card-front"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="mt-10"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="display-2 text--primary"
                    >
                      你好
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
             
            <v-card-stub
              class="mx-auto my-2 pa-2 memorize-card-back"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Nǐ hǎo
                    </p>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <h1
                      class="display-2"
                    >
                      你好
                    </h1>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="mt-3"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Hello
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
          </div>
        </div>
      </v-col-stub>
      <v-col-stub
        lg="3"
        md="4"
        sm="6"
        tag="div"
        xl="2"
      >
        <div
          class="memorize-card"
        >
          <div
            class="memorize-card-inner"
          >
            <v-card-stub
              class="my-2 pa-2 memorize-card-front"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="mt-10"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="display-2 text--primary"
                    >
                      森林
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
             
            <v-card-stub
              class="mx-auto my-2 pa-2 memorize-card-back"
              hover="true"
              loaderheight="4"
              outlined="true"
              tag="div"
            >
              <v-card-text-stub>
                <v-row-stub
                  align="center"
                  justify="center"
                  tag="div"
                >
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Sēnlín
                    </p>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="ma-0 pa-0"
                    cols="12"
                    tag="div"
                  >
                    <h1
                      class="display-2"
                    >
                      森林
                    </h1>
                  </v-col-stub>
                   
                  <v-col-stub
                    class="mt-3"
                    cols="12"
                    tag="div"
                  >
                    <p
                      class="title"
                    >
                      Forest
                    </p>
                  </v-col-stub>
                </v-row-stub>
              </v-card-text-stub>
            </v-card-stub>
          </div>
        </div>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;
